File: C++/search_in_rotated_array.cpp
#include <bits/stdc++.h>
using namespace std;

int Search(vector<int>& arr, int target)
{
    int low = 0;
    int high = arr.size()-1;

    while (low<=high) {

        int mid = (low+high)/2;

        if(arr[mid]==target) return mid;

        if(arr[mid]>=arr[low]){ 

            if(arr[mid]>=target and arr[low]<=target)
                high = mid - 1;
            else low = mid + 1;  

        }  
        else{

            if(arr[mid]<=target and arr[high]>=target) 
                low = mid + 1;
            else high = mid - 1;
            
        }    
        
    }
    return -1;
}

int main()
{
    vector<int> arr = {4,5,6,7,0,1,2};
    int n = sizeof(arr)/sizeof(arr[0]);
    cout << " Search in rotated array " << Search(arr, 0);
    return 0;
}

File: Java/SearchInRotatedArray.java
class SearchInRotatedArray {
    public int search(int[] nums, int target) {

    //     I. Naive Approach | O(n)
    //     for(int i= 0; i< nums.length; i++){
    //         if(nums[i] == target){
    //             return i;
    //         }
    //     }
    //     return -1;
    // }

    // II. Binary Search Appraoch
    int left = 0, right = nums.length -1;
    while(left <= right){
        int mid = (left + right) /2;
        if( nums[mid] == target){
            return mid;
        }
        // is left array sorted?
        if( nums[left] <= nums[mid]){
            if(nums[left] <= target && nums[mid] >= target){
                right = mid - 1;
            }
            else{
                left = mid + 1;
            }
        }
        // no, therefore right array is sorted
        else {
            if(nums[mid] <= target && nums[right] >= target){
                left = mid + 1;
            }
            else{
                right = mid - 1;
            }
        }
    }
    return -1;
}
}